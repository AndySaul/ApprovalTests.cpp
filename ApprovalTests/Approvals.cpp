#include "ApprovalTests/Approvals.h"

namespace ApprovalTests
{

}
