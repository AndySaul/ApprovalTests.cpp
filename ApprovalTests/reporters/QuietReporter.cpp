#include "ApprovalTests/reporters/QuietReporter.h"

namespace ApprovalTests
{

}
