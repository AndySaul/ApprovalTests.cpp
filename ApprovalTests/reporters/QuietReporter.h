#ifndef APPROVALTESTS_CPP_QUIETREPORTER_H
#define APPROVALTESTS_CPP_QUIETREPORTER_H

#include "Reporter.h"
#include <memory>
#include <vector>

// A reporter that does nothing. Failing tests will still fail, but nothing will be launched.
class QuietReporter : public Reporter
{
public:
    bool report(std::string received, std::string approved) const override
    {
        return true;
    }
};

#endif //APPROVALTESTS_CPP_QUIETREPORTER_H
