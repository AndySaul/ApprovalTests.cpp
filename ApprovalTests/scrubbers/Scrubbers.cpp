#include "Scrubbers.h"

#include <string>
#include <functional>
#include <regex>
#include <map>

namespace ApprovalTests
{
    std::string Scrubbers::doNothing(const std::string& input)
    {
        return input;
    }
    std::string Scrubbers::scrubRegex(const std::string& input,
                                      const std::regex& regex,
                                      const Scrubbers::RegexReplacer& replaceFunction)
    {
        std::string result;
        std::string remainder = input;
        std::smatch m;
        while (std::regex_search(remainder, m, regex))
        {
            auto match = m[0];
            auto original_matched_text = match.str();
            auto replacement = replaceFunction(match);
            result += std::string(m.prefix()) + replacement;
            remainder = m.suffix();
        }
        result += remainder;
        return result;
    }

    Scrubber Scrubbers::createRegexScrubber(const std::regex& regexPattern,
                                            const Scrubbers::RegexReplacer& replacer)
    {
        return [=](const std::string& input) {
            return scrubRegex(input, regexPattern, replacer);
        };
    }

    Scrubber Scrubbers::createRegexScrubber(const std::regex& regexPattern,
                                            const std::string& replacementText)
    {
        return createRegexScrubber(regexPattern,
                                   [=](const RegexMatch&) { return replacementText; });
    }

    Scrubber Scrubbers::createRegexScrubber(const std::string& regexString,
                                            const std::string& replacementText)
    {
        return createRegexScrubber(std::regex(regexString), replacementText);
    }

    std::string Scrubbers::scrubGuid(const std::string& input)
    {
        static const std::regex regex("[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-["
                                      "0-9a-fA-F]{4}-[0-9a-fA-F]{12}");

        int matchNumber = 0;
        std::map<std::string, int> matchIndices;
        return scrubRegex(input, regex, [&](const RegexMatch& m) {
            auto guid_match = m.str();

            if (matchIndices[guid_match] == 0)
            {
                matchIndices[guid_match] = ++matchNumber;
            }
            return "guid_" + std::to_string(matchIndices[guid_match]);
        });
    }
}
