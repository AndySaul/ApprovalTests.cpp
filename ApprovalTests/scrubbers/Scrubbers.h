#pragma once

#include <string>

namespace Scrubbers
{
    std::string scrubGuid(const std::string& /*input*/)
    {
        // Match against the input to see if it's a GUID or not
        // If it is, replace with guid_1, return original string
        return "guid_1";
    }
}
