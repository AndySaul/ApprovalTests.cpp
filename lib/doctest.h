#include "doctest.2.4.beta.hpp"
