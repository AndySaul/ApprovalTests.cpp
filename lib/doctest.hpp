#include "doctest.2.3.3.hpp"
