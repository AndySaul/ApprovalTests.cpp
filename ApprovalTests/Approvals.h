#pragma once
#include <string>
#include <functional>
#include <exception>
#include <utility>

#include "ApprovalTests/ApprovalTestsVersion.h"
#include "ApprovalTests/core/FileApprover.h"
#include "reporters/DefaultReporter.h"
#include "reporters/DefaultReporterDisposer.h"
#include "ApprovalTests/core/Reporter.h"
#include "namers/ApprovalTestNamer.h"
#include "writers/ExistingFile.h"
#include "namers/ExistingFileNamer.h"
#include "namers/SubdirectoryDisposer.h"
#include "namers/DefaultNamerDisposer.h"
#include "scrubbers/Scrubbers.h"
#include "core/Options.h"
#include "utilities/MoreHelpMessages.h"

namespace ApprovalTests
{

    class Approvals
    {
    private:
        Approvals() = default;

        ~Approvals() = default;

    public:
        static std::shared_ptr<ApprovalNamer> getDefaultNamer()
        {
            return DefaultNamerFactory::getDefaultNamer()();
        }

        static void verify(const std::string& contents,
                           const Options& options = Options())
        {
            verifyWithExtension(contents, ".txt", options);
        }

        static void verifyWithExtension(const std::string& contents,
                                        const std::string& fileExtensionWithDot,
                                        const Options& options = Options())
        {
            StringWriter writer(options.scrub(contents), fileExtensionWithDot);
            FileApprover::verify(*getDefaultNamer(), writer, options.getReporter());
        }

        // Note that this overload ignores any scrubber in options
        static void verify(const ApprovalWriter& writer,
                           const Options& options = Options())
        {
            FileApprover::verify(*getDefaultNamer(), writer, options.getReporter());
        }

        template <typename T>
        using IsNotDerivedFromWriter =
            typename std::enable_if<!std::is_base_of<ApprovalWriter, T>::value,
                                    int>::type;

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verify(const T& contents, const Options& options = Options())
        {
            verify(StringUtils::toString(contents), options);
        }

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verifyWithExtension(const T& contents,
                                        const std::string& fileExtensionWithDot,
                                        const Options& options = Options())
        {
            verifyWithExtension(
                StringUtils::toString(contents), fileExtensionWithDot, options);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void
        verify(const T& contents, Function converter, const Options& options = Options())
        {
            std::stringstream s;
            converter(contents, s);
            verify(s.str(), options);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void verifyWithExtension(const T& contents,
                                        Function converter,
                                        const std::string& fileExtensionWithDot,
                                        const Options& options = Options())
        {
            std::stringstream s;
            converter(contents, s);
            verifyWithExtension(s.str(), fileExtensionWithDot, options);
        }

        static void
        verifyExceptionMessage(const std::function<void(void)>& functionThatThrows,
                               const Options& options = Options())
        {
            std::string message = "*** no exception thrown ***";
            try
            {
                functionThatThrows();
            }
            catch (const std::exception& e)
            {
                message = e.what();
            }
            verify(message, options);
        }

        template <typename Iterator>
        static void verifyAll(
            const std::string& header,
            const Iterator& start,
            const Iterator& finish,
            std::function<void(typename Iterator::value_type, std::ostream&)> converter,
            const Options& options = Options())
        {
            std::stringstream s;
            if (!header.empty())
            {
                s << header << "\n\n\n";
            }
            for (auto it = start; it != finish; ++it)
            {
                converter(*it, s);
                s << '\n';
            }
            verify(s.str(), options);
        }

        template <typename Container>
        static void verifyAll(
            std::string header,
            const Container& list,
            std::function<void(typename Container::value_type, std::ostream&)> converter,
            const Options& options = Options())
        {
            verifyAll<typename Container::const_iterator>(
                header, list.begin(), list.end(), converter, options);
        }

        template <typename T>
        static void verifyAll(const std::string& header,
                              const std::vector<T>& list,
                              const Options& options = Options())
        {
            int i = 0;
            verifyAll<std::vector<T>>(
                header,
                list,
                [&](T e, std::ostream& s) { s << "[" << i++ << "] = " << e; },
                options);
        }

        template <typename T>
        static void verifyAll(const std::vector<T>& list,
                              const Options& options = Options())
        {
            verifyAll<T>("", list, options);
        }

        // Note that this method ignores any scrubber in options
        static void verifyExistingFile(const std::string& filePath,
                                       const Options& options = Options())
        {
            ExistingFile writer(filePath);
            ExistingFileNamer namer(filePath);
            FileApprover::verify(namer, writer, options.getReporter());
        }

        static SubdirectoryDisposer
        useApprovalsSubdirectory(const std::string& subdirectory = "approval_tests")
        {
            return SubdirectoryDisposer(subdirectory);
        }

        static DefaultReporterDisposer
        useAsDefaultReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return DefaultReporterDisposer(reporter);
        }

        static FrontLoadedReporterDisposer
        useAsFrontLoadedReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return FrontLoadedReporterDisposer(reporter);
        }

        static DefaultNamerDisposer useAsDefaultNamer(NamerCreator namerCreator)
        {
            return DefaultNamerDisposer(std::move(namerCreator));
        }

#ifndef APPROVAL_TESTS_HIDE_DEPRECATED_CODE

        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS
        static void verify(std::string contents, const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verify(contents, Options(reporter));
        }

        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS
        static void verifyWithExtension(std::string contents,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyWithExtension(contents, fileExtensionWithDot, Options(reporter));
        }

        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS
        static void verify(const ApprovalWriter& writer, const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verify(writer, Options(reporter));
        }

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void verify(const T& contents,
                                                                 const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verify(StringUtils::toString(contents), reporter);
        }

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void
        verifyWithExtension(const T& contents,
                            const std::string& fileExtensionWithDot,
                            const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyWithExtension(contents, fileExtensionWithDot, Options(reporter));
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void
        verify(const T& contents, Function converter, const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verify(contents, converter, Options(reporter));
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void
        verifyWithExtension(const T& contents,
                            Function converter,
                            const std::string& fileExtensionWithDot,
                            const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyWithExtension(
                contents, converter, fileExtensionWithDot, Options(reporter));
        }

        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS
        static void verifyExceptionMessage(std::function<void(void)> functionThatThrows,
                                           const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyExceptionMessage(functionThatThrows, Options(reporter));
        }

        template <typename Iterator>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void verifyAll(
            std::string header,
            const Iterator& start,
            const Iterator& finish,
            std::function<void(typename Iterator::value_type, std::ostream&)> converter,
            const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyAll(header, start, finish, converter, Options(reporter));
        }

        template <typename Container>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void verifyAll(
            std::string header,
            const Container& list,
            std::function<void(typename Container::value_type, std::ostream&)> converter,
            const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyAll(header, list, converter, Options(reporter));
        }

        template <typename T>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void verifyAll(
            std::string header, const std::vector<T>& list, const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyAll(header, list, Options(reporter));
        }

        template <typename T>
        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS static void
        verifyAll(const std::vector<T>& list, const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyAll<T>(list, Options(reporter));
        }

        APPROVAL_TESTS_DEPRECATED_USE_OPTIONS
        static void verifyExistingFile(const std::string filePath,
                                       const Reporter& reporter)
        {
            APPROVAL_TESTS_DEPRECATED_USE_OPTIONS_CPP11
            verifyExistingFile(filePath, Options(reporter));
        }
#endif
    };
}
