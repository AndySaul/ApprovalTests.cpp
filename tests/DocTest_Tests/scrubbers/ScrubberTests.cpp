#include "doctest/doctest.h"
#include "ApprovalTests/scrubbers/Scrubbers.h"

#include <iostream> // needed to fix linker error on XCode Release builds
#include "ApprovalTests/Approvals.h"
#include "ApprovalTests/reporters/QuietReporter.h"

#include <cstdlib>
#include <chrono>

using namespace ApprovalTests;

namespace
{
    int random(int max)
    {
        return std::rand() % max;
    }
}

TEST_CASE("test createRegexScrubber")
{
    // begin-snippet: complex_regex_scrubbing
    auto input = "1) Hello 1234 World";
    auto scrubber =
        Scrubbers::createRegexScrubber(std::regex(R"(\d+)"), [](const auto& match) {
            auto match_text = match.str();
            auto match_integer = std::stoi(match_text);
            if (match_integer < 10)
            {
                return match_text;
            }
            else
            {
                return std::string("[number]");
            }
        });
    // end-snippet
    Approvals::verify(input, Options(scrubber));
}

TEST_CASE("test createRegexScrubber with fixed result")
{
    auto input = "Hello 1234 World";
    auto scrubber = Scrubbers::createRegexScrubber(std::regex(R"(\d+)"), "number");
    Approvals::verify(input, Options(scrubber));
}

TEST_CASE("test createRegexScrubber with string input and fixed result")
{
    // begin-snippet: simple_regex_scrubbing
    std::stringstream os;
    os << "Hello " << random(1000) << " World";
    Approvals::verify(os.str(),
                      Options(Scrubbers::createRegexScrubber(R"(\d+)", "[number]")));
    // end-snippet
}

TEST_CASE("regex scrubber")
{
    auto input = "9012 Hello 1234 World 1234 5678";
    auto scrubber =
        Scrubbers::scrubRegex(input, std::regex(R"(\d\d\d\d)"), [](const auto& match) {
            return "number(" + match.str() + ")\n";
        });
    Approvals::verify(scrubber);
}

TEST_CASE("regex scrubber with full customisation")
{
    auto input = "9012 Hello 1234 World 1234 5678";
    Scrubber scrubber = [](const std::string& text) {
        return Scrubbers::scrubRegex(
            text, std::regex(R"(\d\d\d\d)"), [](const auto& match) {
                return "number(" + match.str() + ")\n";
            });
    };
    // begin-snippet: scrubber_in_options_object
    Approvals::verify(input, Options().withScrubber(scrubber));
    // end-snippet
}

TEST_CASE("test createRegexScrubber with date and time")
{
    std::time_t now = std::time(nullptr);

    char mbstr[100];
    if (!std::strftime(mbstr, sizeof(mbstr), "%c", std::localtime(&now)))
    {
        throw std::runtime_error("Error converting date and time to string");
    }

    std::stringstream output;
    output << "Today's date and time is " << mbstr << '\n';

    // begin-snippet: scrubbing_date_and_time
    // Example of format that this matches:
    //  Tue Sep  3 16:58:52 2019
    const auto dateRegex = R"([A-Za-z]{3} [A-Za-z]{3} [\d ]\d \d\d:\d\d:\d\d \d\d\d\d)";

    Approvals::verify(
        output.str(),
        Options(Scrubbers::createRegexScrubber(dateRegex, "[date_and_time]")));
    // end-snippet
}
