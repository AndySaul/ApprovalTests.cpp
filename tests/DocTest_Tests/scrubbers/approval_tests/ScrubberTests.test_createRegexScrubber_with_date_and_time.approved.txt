Today's date and time is [date_and_time]

